### PR TITLE
Add getUserByLoginID admin API

### PR DESCRIPTION
**Pull request by mickmak-busicat**
ref dev-1849

@louischan-oursky check check when you free 🙏

Preview:

https://github.com/user-attachments/assets/88ef08de-ea90-46d6-acd7-bef328d67260

### CODE REVIEW

File: pkg/admin/facade/user.go [hunk 1]
```diff
@@ -33,6 +33,7 @@ type UserService interface {
 	CheckUserAnonymized(userID string) error
 	UpdateMFAEnrollment(userID string, endAt *time.Time) error
 	GetUsersByStandardAttribute(attributeName string, attributeValue string) ([]string, error)
+	GetUsersByLoginID(loginIDKey string, loginIDValue string) (string, error)
```
Comment (louischan-oursky) [inline pkg/admin/facade/user.go:36]: GetUserByLoginID. No s.

File: pkg/admin/graphql/query.go [hunk 2]
```diff
@@ -342,5 +342,30 @@ var query = graphql.NewObject(graphql.ObjectConfig{
 				}), err
 			},
 		},
+		"getUsersByLoginID": &graphql.Field{
```
Comment (louischan-oursky) [inline pkg/admin/graphql/query.go:345]: getUserByLoginID. No s

File: pkg/admin/graphql/query.go [hunk 3]
```diff
@@ -342,5 +342,30 @@ var query = graphql.NewObject(graphql.ObjectConfig{
...
+
+				loginIDKey, _ := p.Args["loginIDKey"].(string)
+				loginIDValue, _ := p.Args["loginIDValue"].(string)
+
+				userID, err := gqlCtx.UserFacade.GetUsersByLoginID(loginIDKey, loginIDValue)
+				if err != nil {
+					return nil, err
+				}
+
+				return graphqlutil.NewLazyValue(gqlCtx.Users.Load(userID)).Value()
```
Comment (louischan-oursky) [inline pkg/admin/graphql/query.go:367]: Maybe you can simplify this with `return gqlCtx.Users.Load(userID).Value()`

File: pkg/lib/authn/identity/service/service.go [hunk 4]
```diff
@@ -1081,3 +1082,7 @@ func (s *Service) listSIWECandidates(siwes []*identity.SIWE) []identity.Candidat
 func (s *Service) Normalize(typ model.LoginIDKeyType, value string) (normalized string, uniqueKey string, err error) {
 	return s.LoginID.Normalize(typ, value)
 }
+
+func (s *Service) GetByKeyAndValue(loginIDKey string, loginIDValue string) (*identity.LoginID, error) {
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/service/service.go:1086]: Service should always return `*identity.Info`. So you need to convert `*identity.LoginID` into `*identity.Info` here. Pay attention to error handling. This method should return `ErrIdentityNotFound` if the value cannot be used to look up an Login ID identity.

File: pkg/lib/facade/coordinator.go [hunk 5]
```diff
@@ -1425,3 +1426,16 @@ func (c *Coordinator) GetUsersByStandardAttribute(attributeName string, attribut
 
 	return uniqueUserIDs, nil
 }
+
+func (c *Coordinator) GetUsersByLoginID(loginIDKey string, loginIDValue string) (string, error) {
+	loginID, err := c.Identities.GetByKeyAndValue(loginIDKey, loginIDValue)
+
+	if errors.Is(err, api.ErrIdentityNotFound) {
+		// User identity not found is not an error in this function.
+		return "", nil
```
Comment (louischan-oursky) [inline pkg/lib/facade/coordinator.go:1436]: You need to return `api.UserNotFound` here.

File: pkg/admin/graphql/query.go [hunk 6]
```diff
@@ -342,5 +342,30 @@ var query = graphql.NewObject(graphql.ObjectConfig{
...
+			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
+				gqlCtx := GQLContext(p.Context)
+
+				loginIDKey, _ := p.Args["loginIDKey"].(string)
+				loginIDValue, _ := p.Args["loginIDValue"].(string)
+
+				userID, err := gqlCtx.UserFacade.GetUsersByLoginID(loginIDKey, loginIDValue)
+				if err != nil {
+					return nil, err
+				}
```
Comment (louischan-oursky) [inline pkg/admin/graphql/query.go:365]: You need to check the error. Specifically, when the err is `UserNotFound`, then you simply `return nil, nil`. Otherwise, you return `gqlCtx.Users.Load(userID).Value()`


```
userID, err := gqlCtx.UserFacade.GetUsersByLoginID(loginIDKey, loginIDValue)
if errors.Is(err, api.ErrUserNotFound) {
  return nil, nil
} else if err != nil {
  return nil, err
}

return gqlCtx.Users.Load(userID).Value()
```

File: pkg/lib/authn/identity/loginid/provider.go [hunk 7]
```diff
@@ -85,6 +85,38 @@ func (p *Provider) GetByValue(value string) ([]*identity.LoginID, error) {
 	return is, nil
 }
 
+func (p *Provider) GetByKeyAndValue(key string, value string) (*identity.LoginID, error) {
+	// Normalize expects loginID is in correct type so we have to validate it first.
+	invalid := p.Checker.ValidateOne(identity.LoginIDSpec{
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/loginid/provider.go:90]: No need to call ValidateOne. What you need to do is

1. Check `key` is a valid login ID key. if not, return `ErrGetUsersInvalidArgument.New("invalid Login ID key")` You do this by calling `LoginIDConfig.GetKeyConfig`
2. Normalize `value` according to the type.

File: pkg/lib/authn/identity/loginid/provider.go [hunk 8]
```diff
@@ -85,6 +85,38 @@ func (p *Provider) GetByValue(value string) ([]*identity.LoginID, error) {
...
+	})
+
+	if invalid != nil {
+		return nil, invalid
+	}
+
+	normalizer := p.NormalizerFactory.NormalizerWithLoginIDType(model.LoginIDKeyType(key))
+	normalizedloginID, err := normalizer.Normalize(value)
+	if err != nil {
+		return nil, err
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/loginid/provider.go]: This is something I missed in the spec. If the value cannot be normalized, then we should return `ErrGetUsersInvalidArgument.New("invalid Login ID value")`. Please update the spec in this PR to reflect this corrected behavior.

File: pkg/lib/facade/coordinator.go [hunk 9]
```diff
@@ -1433,6 +1434,10 @@ func (c *Coordinator) GetUsersByLoginID(loginIDKey string, loginIDValue string)
 	if errors.Is(err, api.ErrIdentityNotFound) {
 		// User identity not found is not an error in this function.
 		return "", nil
+	} else if errors.Is(err, loginid.ErrNormalize) {
+		return "", api.ErrGetUsersInvalidArgument.New("invalid Login ID value")
+	} else if errors.Is(err, loginid.ErrValidate) {
+		return "", api.ErrGetUsersInvalidArgument.New("invalid Login ID key")
```
Comment (louischan-oursky) [inline pkg/lib/facade/coordinator.go:1440]: With my proposed change, we do not need this commit.

File: e2e/tests/admin_api/get_user_by_loginid_test.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+		})
+		So(err, ShouldBeNil)
+		So(rawResp, ShouldEqualJSON, `
+			{
+				"data": {
+					"getUsersByLoginID": null
+				},
+				"errors": [
+					{
+						"message": "invalid Login ID key",
```
Comment (louischan-oursky) [inline e2e/tests/admin_api/get_user_by_loginid_test.go:217]: The message is wrong. It should be `invalid Login ID value`

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: e2e/tests/admin_api/get_user_by_loginid_test.go (hunk 10 above)
Comment (mickmak-busicat) [reply e2e/tests/admin_api/get_user_by_loginid_test.go]: fixed e2e test

File: pkg/lib/facade/coordinator.go (hunk 9 above)
Comment (mickmak-busicat) [reply pkg/lib/facade/coordinator.go]: removed all error handling related to `loginid.ErrValidate`

File: pkg/lib/authn/identity/loginid/provider.go (hunk 8 above)
Comment (mickmak-busicat) [reply pkg/lib/authn/identity/loginid/provider.go]: updated error handling in this function

File: pkg/lib/authn/identity/loginid/provider.go (hunk 7 above)
Comment (mickmak-busicat) [reply pkg/lib/authn/identity/loginid/provider.go]: updated to use `LoginIDConfig`

File: pkg/admin/graphql/query.go (hunk 6 above)
Comment (mickmak-busicat) [reply pkg/admin/graphql/query.go:370]: Updated to handle `api.ErrUserNotFound`

File: pkg/lib/facade/coordinator.go (hunk 5 above)
Comment (mickmak-busicat) [reply pkg/lib/facade/coordinator.go]: updated to return `api.ErrUserNotFound`

File: pkg/admin/graphql/query.go (hunk 3 above)
Comment (mickmak-busicat) [reply pkg/admin/graphql/query.go]: fixed no need `NewLazyValue`

File: pkg/admin/graphql/query.go (hunk 2 above)
Comment (mickmak-busicat) [reply pkg/admin/graphql/query.go]: fixed to `getUserByLoginID`

File: pkg/admin/facade/user.go (hunk 1 above)
Comment (mickmak-busicat) [reply pkg/admin/facade/user.go]: fixed to `getUserByLoginID`

**Comment by mickmak-busicat** [comment]
@louischan-oursky fixed problem in comments 🙏

File: pkg/lib/authn/identity/loginid/errors.go [hunk 11]
```diff
@@ -4,5 +4,4 @@ import (
 	"errors"
 )
 
-var ErrValidate = errors.New("ValidateError")
 var ErrNormalize = errors.New("NormalizationError")
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/loginid/errors.go:7]: Please also remove this newly introduced error. We no longer need it.

File: pkg/lib/authn/identity/loginid/provider.go [hunk 12]
```diff
@@ -86,23 +86,13 @@ func (p *Provider) GetByValue(value string) ([]*identity.LoginID, error) {
...
-		return nil, errors.Join(ErrValidate, invalid)
+	if !ok {
+		return nil, api.ErrGetUsersInvalidArgument.New("invalid Login ID key")
 	}
 
-	normalizer := p.NormalizerFactory.NormalizerWithLoginIDType(model.LoginIDKeyType(key))
+	normalizer := p.NormalizerFactory.NormalizerWithLoginIDType(cfg.Type)
 	normalizedloginID, err := normalizer.Normalize(value)
 	if err != nil {
 		return nil, errors.Join(ErrNormalize, err)
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/loginid/provider.go:98]: You return `api.ErrGetUsersInvalidArgument.New("invalid Login ID value"` here. Because you know exactly the error is due to failed normalization.

File: pkg/lib/authn/identity/loginid/provider.go [hunk 13]
```diff
@@ -65,7 +65,7 @@ func (p *Provider) GetByValue(value string) ([]*identity.LoginID, error) {
 		normalizer := p.NormalizerFactory.NormalizerWithLoginIDType(config.Type)
 		normalizedloginID, err := normalizer.Normalize(value)
 		if err != nil {
-			return nil, err
+			return nil, errors.Join(ErrNormalize, err)
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/loginid/provider.go:68]: Revert this change.

File: pkg/lib/authn/identity/loginid/provider.go [hunk 14]
```diff
@@ -139,7 +161,7 @@ func (p *Provider) New(userID string, spec identity.LoginIDSpec, options Checker
 
 	normalized, uniqueKey, err := p.Normalize(spec.Type, spec.Value)
 	if err != nil {
-		return nil, err
+		return nil, errors.Join(ErrNormalize, err)
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/loginid/provider.go:164]: Revert this change.

File: pkg/lib/authn/identity/loginid/provider.go [hunk 15]
```diff
@@ -175,7 +197,7 @@ func (p *Provider) WithValue(iden *identity.LoginID, value string, options Check
...
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/loginid/provider.go:200]: Revert this change.

File: pkg/lib/facade/coordinator.go [hunk 16]
```diff
@@ -1425,3 +1427,17 @@ func (c *Coordinator) GetUsersByStandardAttribute(attributeName string, attribut
 
 	return uniqueUserIDs, nil
 }
+
+func (c *Coordinator) GetUserByLoginID(loginIDKey string, loginIDValue string) (string, error) {
+	info, err := c.Identities.AdminAPIGetByLoginIDKeyAndLoginIDValue(loginIDKey, loginIDValue)
+
+	if errors.Is(err, api.ErrIdentityNotFound) {
+		return "", api.ErrUserNotFound
+	} else if errors.Is(err, loginid.ErrNormalize) {
```
Comment (louischan-oursky) [inline pkg/lib/facade/coordinator.go:1436]: Since ErrNormalize is removed, and the error is constructed downstream, we do not need this if statement checking.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/lib/authn/identity/loginid/errors.go (hunk 11 above)
Comment (mickmak-busicat) [reply pkg/lib/authn/identity/loginid/errors.go:7]: Removed `ErrNormalize` related code

File: pkg/lib/authn/identity/loginid/provider.go (hunk 12 above)
Comment (mickmak-busicat) [reply pkg/lib/authn/identity/loginid/provider.go:98]: Removed `ErrNormalize` related code

File: pkg/lib/authn/identity/loginid/provider.go (hunk 13 above)
Comment (mickmak-busicat) [reply pkg/lib/authn/identity/loginid/provider.go:68]: Removed `ErrNormalize` related code

File: pkg/lib/authn/identity/loginid/provider.go (hunk 14 above)
Comment (mickmak-busicat) [reply pkg/lib/authn/identity/loginid/provider.go:164]: Removed `ErrNormalize` related code

File: pkg/lib/authn/identity/loginid/provider.go (hunk 15 above)
Comment (mickmak-busicat) [reply pkg/lib/authn/identity/loginid/provider.go:200]: Removed `ErrNormalize` related code

File: pkg/lib/facade/coordinator.go (hunk 16 above)
Comment (mickmak-busicat) [reply pkg/lib/facade/coordinator.go:1436]: Removed `ErrNormalize` related code

**Comment by mickmak-busicat** [comment]
@louischan-oursky removed `ErrNormalize` related logic 🙏

**Review by louischan-oursky — APPROVED** [review]
